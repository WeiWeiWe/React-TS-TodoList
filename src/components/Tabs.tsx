import React from 'react';
import styled from 'styled-components';

const TabsStyle = styled.section`
  border-bottom: 1px solid #bbb;
  .tab-fields {
    display: flex;
    justify-content: flex-end;
    height: 100%;
    ul {
      display: flex;
      align-items: center;

      li {
        display: flex;
        align-items: center;
        margin: 10px;
        padding: 10px;
        cursor: pointer;

        h2 {
          color: rgb(50, 197, 250);
          font-weight: bold;
        }
      }
    }
  }
`;

function Tabs() {
  return (
    <TabsStyle className="section-fields">
      <nav className="tab-fields">
        <ul>
          <li>
            <h2>ACTIVE</h2>
          </li>
          <li>
            <h2>DONE</h2>
          </li>
          <li>
            <h2>ALL</h2>
          </li>
        </ul>
      </nav>
    </TabsStyle>
  );
}

export default Tabs;
