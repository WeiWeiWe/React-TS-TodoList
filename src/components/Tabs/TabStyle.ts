import styled from 'styled-components';

export const TabsStyle = styled.section`
  border-bottom: 1px solid #bbb;
  .tab-fields {
    display: flex;
    justify-content: flex-end;
    height: 100%;
    button {
      display: flex;
      align-items: center;
      margin: 10px;
      padding: 10px;
      background: #fff;
      border: none;
      cursor: pointer;

      &:hover {
        background: rgba(50, 197, 250, 0.2);
      }
      h2 {
        color: rgb(50, 197, 250);
        font-weight: bold;
        font-size: 18px;
      }
    }
    .isClick {
      background: rgb(50, 197, 250, 0.5);

      h2 {
        color: #fff;
      }
    }
  }
`;
