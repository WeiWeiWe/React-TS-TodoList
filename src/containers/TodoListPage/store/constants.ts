export const CHANGE_INPUT_VALUE = 'CHANGE_INPUT_VALUE';
export const ADD_TODO_LIST = 'ADD_TODO_LIST';
export const DELETE_TODO_LIST = 'DELETE_TODO_LIST';
export const CHANGE_LIST_ITEM_TITLE_VALUE = 'CHANGE_LIST_ITEM_TITLE_VALUE';
export const CHANGE_PROCESS_STATUS = 'CHANGE_PROCESS_STATUS';
export const CHANGE_TAB_STATUS = 'CHANGE_TAB_STATUS';
