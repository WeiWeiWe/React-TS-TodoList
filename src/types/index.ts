export * from './Type';
export * from './Enum';
