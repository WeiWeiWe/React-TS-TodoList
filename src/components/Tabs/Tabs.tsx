import React from 'react';
import { TabsStyle } from './TabStyle';
import { EnumItemProcessTypes, ItemTypes } from '../../types';

export interface TabsIprops {
  changeTabStatus: (e: React.MouseEvent<HTMLButtonElement>) => void;
  list: ItemTypes[];
  tabStatus: EnumItemProcessTypes;
}

export function Tabs({ changeTabStatus, list, tabStatus }: TabsIprops) {
  return (
    <TabsStyle>
      <nav className="global-block-fields tabs-fields">
        <button
          className={
            list?.length > 0
              ? tabStatus === EnumItemProcessTypes.ACTIVE
                ? 'tabs-button isClick'
                : 'tabs-button'
              : 'tabs-button'
          }
          value={EnumItemProcessTypes.ACTIVE}
          onClick={changeTabStatus}
          data-testid="tabs-active-btn"
        >
          <h2>ACTIVE</h2>
        </button>
        <button
          className={
            list?.length > 0
              ? tabStatus === EnumItemProcessTypes.DONE
                ? 'tabs-button isClick'
                : 'tabs-button'
              : 'tabs-button'
          }
          value={EnumItemProcessTypes.DONE}
          onClick={changeTabStatus}
          data-testid="tabs-done-btn"
        >
          <h2>DONE</h2>
        </button>
        <button
          className={
            list?.length > 0
              ? tabStatus === EnumItemProcessTypes.All
                ? 'tabs-button isClick'
                : 'tabs-button'
              : 'tabs-button'
          }
          value={EnumItemProcessTypes.All}
          onClick={changeTabStatus}
          data-testid="tabs-all-btn"
        >
          <h2>ALL</h2>
        </button>
      </nav>
    </TabsStyle>
  );
}
