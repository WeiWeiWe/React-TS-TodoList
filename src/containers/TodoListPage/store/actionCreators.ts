import * as constants from './constants';
import { ItemTypes, EnumItemProcessTypes } from '../../../types';

interface ChangeInputValueAction {
  type: typeof constants.CHANGE_INPUT_VALUE;
  value: string;
}

interface AddTodoListAction {
  type: typeof constants.ADD_TODO_LIST;
  item: ItemTypes;
}

interface DeleteTodoListAction {
  type: typeof constants.DELETE_TODO_LIST;
  id: string;
}

interface SwitchModifyStatusAction {
  type: typeof constants.SWITCH_MODIFY_STATUS;
  status: boolean;
  id: string;
}

interface ChangeListItemTitleValueAction {
  type: typeof constants.CHANGE_LIST_ITEM_TITLE_VALUE;
  value: string;
  id: string;
}

interface ChangeProcessStatusAction {
  type: typeof constants.CHANGE_PROCESS_STATUS;
  process: string;
  id: string;
}

interface ChangeTabStatusAction {
  type: typeof constants.CHANGE_TAB_STATUS;
  status: EnumItemProcessTypes;
}

export type TodoListActionTypes =
  | ChangeInputValueAction
  | AddTodoListAction
  | DeleteTodoListAction
  | SwitchModifyStatusAction
  | ChangeListItemTitleValueAction
  | ChangeProcessStatusAction
  | ChangeTabStatusAction;

export const changeInputValueActionCreator = (value: string): ChangeInputValueAction => {
  return {
    type: constants.CHANGE_INPUT_VALUE,
    value,
  };
};

export const addTodoListActionCreator = (item: ItemTypes): AddTodoListAction => {
  return {
    type: constants.ADD_TODO_LIST,
    item,
  };
};

export const deleteTodoListActionCreator = (id: string): DeleteTodoListAction => {
  return {
    type: constants.DELETE_TODO_LIST,
    id,
  };
};

export const switchModifyStatusActionCreator = (status: boolean, id: string): SwitchModifyStatusAction => {
  return {
    type: constants.SWITCH_MODIFY_STATUS,
    status,
    id,
  };
};

export const ChangeListItemTitleValueActionCreator = (value: string, id: string): ChangeListItemTitleValueAction => {
  return {
    type: constants.CHANGE_LIST_ITEM_TITLE_VALUE,
    value,
    id,
  };
};

export const ChangeProcessStatusActionCreator = (process: string, id: string): ChangeProcessStatusAction => {
  return {
    type: constants.CHANGE_PROCESS_STATUS,
    process,
    id,
  };
};

export const ChangeTabStatusActionCreator = (status: EnumItemProcessTypes): ChangeTabStatusAction => {
  return {
    type: constants.CHANGE_TAB_STATUS,
    status,
  };
};
