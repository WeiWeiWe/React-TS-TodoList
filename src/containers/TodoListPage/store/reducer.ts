import * as constants from './constants';
import * as actionCreators from './actionCreators';
import { ItemTypes } from '../../../types';

export interface defaultStateTypes {
  inputValue: string;
  list: ItemTypes[];
}

const defaultState: defaultStateTypes = {
  inputValue: '',
  list: [],
};

export default (state = defaultState, action: actionCreators.TodoListActionTypes) => {
  switch (action.type) {
    case constants.CHANGE_INPUT_VALUE:
      return {
        ...state,
        inputValue: action.value,
      };
    case constants.ADD_TODO_LIST:
      return {
        ...state,
        inputValue: '',
        list: [...state.list, action.item],
      };
    case constants.DELETE_TODO_LIST:
      const newList = state.list.filter((item, index) => index !== action.index);
      return {
        ...state,
        list: newList,
      };
    default:
      return state;
  }
};
