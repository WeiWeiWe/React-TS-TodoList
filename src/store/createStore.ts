import { createStore, combineReducers } from 'redux';
import { reducer as todoReducer } from '../containers/TodoListPage/store';

const rootReducer = combineReducers({
  todo: todoReducer,
});

const store = createStore(rootReducer);

export type RootState = ReturnType<typeof store.getState>;

export default store;
