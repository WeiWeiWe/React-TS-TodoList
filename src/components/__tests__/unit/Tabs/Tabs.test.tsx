import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import store from '../../../../store/createStore';
import { Tabs, TabsIprops } from '../../../../components/Tabs/Tabs';
import { EnumItemProcessTypes } from '../../../../types';

const testProps: TabsIprops = {
  changeTabStatus: jest.fn(),
  list: [{ id: '0', title: 'Buy pencil', process: EnumItemProcessTypes.ACTIVE, isModify: false }],
  tabStatus: EnumItemProcessTypes.All,
};

const renderComponentWithRedux = (testProps: TabsIprops) => {
  return (
    <Provider store={store}>
      <Tabs {...testProps} />
    </Provider>
  );
};

describe('Test Tabs Component', () => {
  test('should render the component to match the snapshot', () => {
    const wrapper = render(renderComponentWithRedux(testProps));
    expect(wrapper).toMatchSnapshot();
  });

  test('When user click the all button, should add isClick class name', () => {
    const { getByTestId } = render(renderComponentWithRedux({ ...testProps, tabStatus: EnumItemProcessTypes.All }));
    const allButton = getByTestId(/tabs-all-btn/);

    fireEvent.click(allButton);
    expect(testProps.changeTabStatus).toHaveBeenCalled();
    expect(allButton).toHaveClass('tabs-button isClick');
  });

  test('When user click the active button, should add isClick class name', () => {
    const { getByTestId } = render(renderComponentWithRedux({ ...testProps, tabStatus: EnumItemProcessTypes.ACTIVE }));
    const activeButton = getByTestId(/tabs-active-btn/);

    fireEvent.click(activeButton);
    expect(testProps.changeTabStatus).toHaveBeenCalled();
    expect(activeButton).toHaveClass('tabs-button isClick');
  });

  test('When user click the done button, should add isClick class name', () => {
    const { getByTestId } = render(renderComponentWithRedux({ ...testProps, tabStatus: EnumItemProcessTypes.DONE }));
    const doneButton = getByTestId(/tabs-done-btn/);

    fireEvent.click(doneButton);
    expect(testProps.changeTabStatus).toHaveBeenCalled();
    expect(doneButton).toHaveClass('tabs-button isClick');
  });
});
