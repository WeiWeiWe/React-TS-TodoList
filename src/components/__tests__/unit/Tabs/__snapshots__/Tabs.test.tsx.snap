// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tabs Component should render the component to match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="sc-bdfBQB hrmgu"
      >
        <nav
          class="global-block-fields tabs-fields"
        >
          <button
            class="tabs-button"
            data-testid="active-btn"
            value="ACTIVE"
          >
            <h2>
              ACTIVE
            </h2>
          </button>
          <button
            class="tabs-button"
            data-testid="done-btn"
            value="DONE"
          >
            <h2>
              DONE
            </h2>
          </button>
          <button
            class="tabs-button isClick"
            data-testid="all-btn"
            value="ALL"
          >
            <h2>
              ALL
            </h2>
          </button>
        </nav>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="sc-bdfBQB hrmgu"
    >
      <nav
        class="global-block-fields tabs-fields"
      >
        <button
          class="tabs-button"
          data-testid="active-btn"
          value="ACTIVE"
        >
          <h2>
            ACTIVE
          </h2>
        </button>
        <button
          class="tabs-button"
          data-testid="done-btn"
          value="DONE"
        >
          <h2>
            DONE
          </h2>
        </button>
        <button
          class="tabs-button isClick"
          data-testid="all-btn"
          value="ALL"
        >
          <h2>
            ALL
          </h2>
        </button>
      </nav>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
